File: src/session/learning-viewport.ts
/**
 * Learning viewport builder — projects convention bundle and module internals
 * into viewport response types for UI consumption.
 *
 * Lives in session/ because it reads from conventions/ (ConventionBundle,
 * ConventionModule) and writes to service response types. The UI never
 * imports this file — it calls service methods.
 */

import type { Call } from "../engine/types";
import type { ConventionModule, LocalFsm, ExplanationEntry, BidMeaningClause, SystemConfig, ObsPattern } from "../conventions";
import {
  moduleSurfaces, getModule, getAllModules, listBundleInputs,
  AVAILABLE_BASE_SYSTEMS, deriveNeutralDescription,
  getBaseModuleIds, getSystemConfig, normalizeIntent, matchObs,
  getPrimaryCapability,
} from "../conventions";
import { formatTransitionLabel } from "./format-obs-label";
import { parsePatternCall } from "../engine/auction-helpers";
import { formatCall, formatBidReferences } from "../service/display/format";
import type {
  ModuleCatalogEntry,
  ModuleLearningViewport,
  PhaseGroupView,
  SurfaceDetailView,
  SurfaceClauseView,
  ClauseSystemVariant,
  BaseModuleInfo,
} from "../service/response-types";
import type { BaseSystemId } from "../conventions";

/** Known bridge abbreviations that should be fully uppercased. */
const BRIDGE_ABBREVIATIONS = new Set(["nt", "sayc", "hcp"]);

/** Convert kebab-case module ID to display name. */
export function formatModuleName(moduleId: string): string {
  if (moduleId === "") return "";
  return moduleId
    .split("-")
    .map((w) => {
      const lower = w.toLowerCase();
      if (BRIDGE_ABBREVIATIONS.has(lower)) return w.toUpperCase();
      const match = lower.match(/^(\d+)(.+)$/);
      if (match && BRIDGE_ABBREVIATIONS.has(match[2]!)) {
        return match[1] + match[2]!.toUpperCase();
      }
      return w.charAt(0).toUpperCase() + w.slice(1);
    })
    .join(" ");
}

// ── Module-centric viewport ──────────────────────────────────────────

// ── System-fact clause helpers ───────────────────────────────────────

/** All system configs, paired with short UI labels. */
const ALL_SYSTEMS: readonly { sys: SystemConfig; label: string }[] =
  AVAILABLE_BASE_SYSTEMS.map((meta) => ({
    sys: getSystemConfig(meta.id),
    label: meta.shortLabel,
  }));

/**
 * Describe what a system-derived fact concretely means for a given SystemConfig.
 * Returns the HCP description + optional trump/NT total-point descriptions, or null for unrecognized facts.
 */
function describeSystemFactValue(
  factId: string,
  sys: SystemConfig,
): { hcp: string; trumpTp?: string } | null {
  switch (factId) {
    case "system.responder.weakHand": {
      const t = sys.responderThresholds;
      return { hcp: `< ${t.inviteMin}`, trumpTp: `< ${t.inviteMinTp.trump}` };
    }
    case "system.responder.inviteValues": {
      const t = sys.responderThresholds;
      return {
        hcp: `${t.inviteMin}\u2013${t.inviteMax}`,
        trumpTp: `${t.inviteMinTp.trump}\u2013${t.inviteMaxTp.trump}`,
      };
    }
    case "system.responder.gameValues": {
      const t = sys.responderThresholds;
      return { hcp: `${t.gameMin}+`, trumpTp: `${t.gameMinTp.trump}+` };
    }
    case "system.responder.slamValues": {
      const t = sys.responderThresholds;
      return { hcp: `${t.slamMin}+`, trumpTp: `${t.slamMinTp.trump}+` };
    }
    case "system.opener.notMinimum": {
      const r = sys.openerRebid;
      return { hcp: `${r.notMinimum}+`, trumpTp: `${r.notMinimumTp.trump}+` };
    }
    case "system.responder.twoLevelNewSuit":
      return { hcp: `${sys.suitResponse.twoLevelMin}+ HCP` };
    case "system.suitResponse.isGameForcing":
      return { hcp: sys.suitResponse.twoLevelForcingDuration === "game" ? "Game-forcing" : "One-round forcing" };
    case "system.oneNtResponseAfterMajor.forcing":
      return { hcp: `1NT is ${sys.oneNtResponseAfterMajor.forcing}` };
    case "system.responder.oneNtRange":
      return { hcp: `${sys.oneNtResponseAfterMajor.minHcp}\u2013${sys.oneNtResponseAfterMajor.maxHcp} HCP` };
    case "system.dontOvercall.inRange":
      return { hcp: `${sys.dontOvercall.minHcp}\u2013${sys.dontOvercall.maxHcp} HCP` };
    default:
      return null;
  }
}

/** Build system variants for a system.* fact — one entry per known base system. */
function buildSystemVariants(factId: string): readonly ClauseSystemVariant[] {
  return ALL_SYSTEMS.map(({ sys, label }) => {
    const result = describeSystemFactValue(factId, sys);
    if (!result) return { systemLabel: label, description: factId };
    return {
      systemLabel: label,
      description: result.hcp,
      trumpTpDescription: result.trumpTp,
    };
  });
}

/** Read the description runtime-property from a BidMeaningClause (set by createSurface). */
function readClauseDescription(c: BidMeaningClause): string {
  return (c as BidMeaningClause & { description?: string }).description
    ?? `${c.factId} ${c.operator} ${JSON.stringify(c.value)}`;
}

// ── Module catalog ──────────────────────────────────────────────────

/** Build module catalog entries for all registered modules. */
export function buildModuleCatalog(): readonly ModuleCatalogEntry[] {
  const allModules = getAllModules();
  const bundleInputs = listBundleInputs();

  // Build reverse map: moduleId → bundleIds that contain it
  const moduleBundles = new Map<string, string[]>();
  for (const input of bundleInputs) {
    for (const memberId of input.memberIds) {
      const existing = moduleBundles.get(memberId);
      if (existing) existing.push(input.id);
      else moduleBundles.set(memberId, [input.id]);
    }
  }

  return allModules.map((mod) => ({
    moduleId: mod.moduleId,
    displayName: formatModuleName(mod.moduleId),
    description: formatBidReferences(mod.description),
    purpose: formatBidReferences(mod.purpose),
    surfaceCount: moduleSurfaces(mod).length,
    bundleIds: moduleBundles.get(mod.moduleId) ?? [],
  }));
}

/** Build read-only metadata for base system modules (for settings display). */
export function buildBaseModuleInfos(baseSystemId: BaseSystemId): readonly BaseModuleInfo[] {
  const ids = getBaseModuleIds(baseSystemId);
  return ids.map((id) => {
    const mod = getModule(id);
    return {
      id,
      displayName: formatModuleName(id),
      description: mod?.description ?? id,
    };
  });
}

/** Build a full learning viewport for a single module. */
export function buildModuleLearningViewport(moduleId: string): ModuleLearningViewport | null {
  const mod = getModule(moduleId);
  if (!mod) return null;

  const bundleInputs = listBundleInputs();
  const bundleIds = bundleInputs
    .filter((b) => b.memberIds.includes(moduleId))
    .map((b) => b.id);

  const teaching = mod.teaching;
  const phases = buildPhaseGroups(mod);

  return {
    moduleId: mod.moduleId,
    displayName: formatModuleName(mod.moduleId),
    description: formatBidReferences(mod.description),
    purpose: formatBidReferences(mod.purpose),
    teaching: {
      tradeoff: teaching?.tradeoff ? formatBidReferences(teaching.tradeoff) : null,
      principle: teaching?.principle ? formatBidReferences(teaching.principle) : null,
      commonMistakes: (teaching?.commonMistakes ?? []).map(formatBidReferences),
    },
    phases,
    bundleIds,
  };
}

// ── Phase grouping ──────────────────────────────────────────────────

/**
 * Derive topological phase order from LocalFsm transitions.
 * Walks from initial state through transitions to produce ordered phases.
 */
export function derivePhaseOrder(fsm: LocalFsm): string[] {
  const phases: string[] = [fsm.initial];
  const seen = new Set<string>([fsm.initial]);

  // BFS from initial following transition edges
  const adjacency = new Map<string, string[]>();
  for (const t of fsm.transitions) {
    const froms: readonly string[] = Array.isArray(t.from) ? t.from : [t.from];
    for (const f of froms) {
      const existing = adjacency.get(f);
      if (existing) {
        if (!existing.includes(t.to)) existing.push(t.to);
      } else {
        adjacency.set(f, [t.to]);
      }
    }
  }

  const queue = [fsm.initial];
  while (queue.length > 0) {
    const current = queue.shift()!;
    const neighbors = adjacency.get(current) ?? [];
    for (const next of neighbors) {
      if (!seen.has(next)) {
        seen.add(next);
        phases.push(next);
        queue.push(next);
      }
    }
  }

  return phases;
}

/**
 * Compute the set of post-fit phases for a module.
 * A phase is post-fit if any StateEntry at that phase has `negotiationDelta.fitAgreed` truthy,
 * or if it's reachable downstream from such a phase via FSM transitions.
 */
export function computePostFitPhases(mod: ConventionModule): Set<string> {
  const fitPhases = new Set<string>();
  for (const entry of mod.states ?? []) {
    if (entry.negotiationDelta?.fitAgreed) {
      const phases: readonly string[] = Array.isArray(entry.phase) ? entry.phase : [entry.phase];
      for (const p of phases) fitPhases.add(p);
    }
  }

  // Build adjacency map from transitions
  const adjacency = new Map<string, string[]>();
  for (const t of mod.local.transitions) {
    const froms: readonly string[] = Array.isArray(t.from) ? t.from : [t.from];
    for (const f of froms) {
      const existing = adjacency.get(f);
      if (existing) {
        if (!existing.includes(t.to)) existing.push(t.to);
      } else {
        adjacency.set(f, [t.to]);
      }
    }
  }

  // BFS from fit-establishing phases to find all reachable downstream phases
  const result = new Set(fitPhases);
  const queue = [...fitPhases];
  while (queue.length > 0) {
    const current = queue.shift()!;
    for (const next of adjacency.get(current) ?? []) {
      if (!result.has(next)) {
        result.add(next);
        queue.push(next);
      }
    }
  }

  return result;
}

/** Format a phase string for display. "asked" → "Asked", "shown-hearts" → "Shown Hearts". */
function formatPhaseDisplay(phase: string, turn: string | null): string {
  const phaseTitle = formatModuleName(phase);
  if (turn) {
    const turnTitle = turn.charAt(0).toUpperCase() + turn.slice(1);
    return `${phaseTitle} — ${turnTitle}`;
  }
  return phaseTitle;
}

/** Find explanation text for a meaningId from module explanation entries. */
export function findExplanationText(entries: readonly ExplanationEntry[], meaningId: string): string | null {
  for (const entry of entries) {
    if ("meaningId" in entry && entry.meaningId === meaningId) {
      return entry.displayText;
    }
  }
  return null;
}

/** Map raw BidMeaningClause[] to SurfaceClauseView[] for the learning viewport.
 *  system.* facts automatically get neutral descriptions + per-system variants.
 *  When `metric` is provided, system-fact clauses get a `relevantMetric` field. */
export function mapClauses(
  clauses: readonly BidMeaningClause[],
  metric?: "hcp" | "trumpTp",
): readonly SurfaceClauseView[] {
  return clauses.map((c) => {
    const isSystemFact = c.factId.startsWith("system.");
    return {
      factId: c.factId,
      operator: c.operator as unknown as SurfaceClauseView["operator"],
      value: c.value,
      description: isSystemFact
        ? deriveNeutralDescription(c.factId, c.rationale)
        : readClauseDescription(c),
      isPublic: c.isPublic ?? false,
      ...(isSystemFact ? { systemVariants: buildSystemVariants(c.factId), relevantMetric: metric } : {}),
    };
  });
}

/** Entry condition info for a module's root: label + optional trigger call. */
export interface EntryCondition {
  readonly label: string;
  readonly call: Call | null;
  readonly turn: "opener" | "responder" | null;
}

const CAP_ENTRY_CONDITIONS: Record<string, EntryCondition> = {
  "opening.1nt": { label: "Partner opened 1NT", call: parsePatternCall("1NT"), turn: "opener" },
  "opening.major": { label: "Partner opened a major", call: null, turn: "opener" },
  "opening.weak-two": { label: "Partner opened a weak two", call: null, turn: "opener" },
  "opponent.1nt": { label: "Opponent opened 1NT", call: parsePatternCall("1NT"), turn: null },
};

/**
 * Derive root phase label from the module's host capability.
 */
function deriveRootPhaseLabel(moduleId: string): string | null {
  const entry = deriveEntryCondition(moduleId);
  return entry?.label ?? null;
}

/** Derive full entry condition (label + trigger call) from the module's host capability. */
export function deriveEntryCondition(moduleId: string): EntryCondition | null {
  for (const input of listBundleInputs()) {
    if (!input.memberIds.includes(moduleId)) continue;
    const capId = getPrimaryCapability(input.declaredCapabilities);
    if (capId && CAP_ENTRY_CONDITIONS[capId]) return CAP_ENTRY_CONDITIONS[capId];
  }
  return null;
}

function findTriggerCall(mod: ConventionModule, fromPhase: string, obs: ObsPattern): Call | null {
  for (const entry of mod.states ?? []) {
    const entryPhases: readonly string[] = Array.isArray(entry.phase) ? entry.phase : [entry.phase];
    if (!entryPhases.includes(fromPhase)) continue;
    for (const surface of entry.surfaces) {
      const actions = normalizeIntent(surface.sourceIntent);
      if (actions.some((a) => matchObs(obs, a))) return surface.encoding.defaultCall;
    }
  }
  return null;
}

/** Build PhaseGroupView[] from a module's states, ordered by FSM topology. */
function buildPhaseGroups(mod: ConventionModule): readonly PhaseGroupView[] {
  const states = mod.states ?? [];
  if (states.length === 0) return [];

  const phaseOrder = derivePhaseOrder(mod.local);
  const postFitPhases = computePostFitPhases(mod);

  const incomingMap = new Map<string, { obs: ObsPattern; fromPhase: string }[]>();
  for (const t of mod.local.transitions) {
    const froms: readonly string[] = Array.isArray(t.from) ? t.from : [t.from];
    for (const f of froms) {
      const existing = incomingMap.get(t.to) ?? [];
      existing.push({ obs: t.on, fromPhase: f });
      incomingMap.set(t.to, existing);
    }
  }

  // Group states by phase string (flatten multi-phase entries)
  const phaseMap = new Map<string, { turn: string | null; surfaces: SurfaceDetailView[] }>();

  for (const entry of states) {
    const entryPhases: readonly string[] = Array.isArray(entry.phase)
      ? entry.phase
      : [entry.phase];

    for (const phase of entryPhases) {
      let group = phaseMap.get(phase);
      if (!group) {
        group = { turn: entry.turn ?? null, surfaces: [] };
        phaseMap.set(phase, group);
      }

      const metric = postFitPhases.has(phase) ? "trumpTp" as const : "hcp" as const;
      const seen = new Set(group.surfaces.map((s) => s.meaningId));
      for (const surface of entry.surfaces) {
        if (seen.has(surface.meaningId)) continue;
        seen.add(surface.meaningId);

        const rawExplanation = findExplanationText(mod.explanationEntries, surface.meaningId);
        group.surfaces.push({
          meaningId: surface.meaningId,
          teachingLabel: {
            name: formatBidReferences(surface.teachingLabel.name),
            summary: formatBidReferences(surface.teachingLabel.summary),
          },
          call: surface.encoding.defaultCall,
          callDisplay: formatCall(surface.encoding.defaultCall),
          disclosure: surface.disclosure,
          recommendation: surface.ranking.recommendationBand ?? null,
          explanationText: rawExplanation ? formatBidReferences(rawExplanation) : null,
          clauses: mapClauses(surface.clauses, metric),
        });
      }
    }
  }

  const visiblePhases = phaseOrder.filter((p) => {
    const g = phaseMap.get(p);
    return g && g.surfaces.length > 0;
  });
  const suppressLabels = visiblePhases.length < 3;

  const result: PhaseGroupView[] = [];
  for (const phase of phaseOrder) {
    const group = phaseMap.get(phase);
    if (!group || group.surfaces.length === 0) continue;

    let transitionLabel: string | null = null;
    if (!suppressLabels) {
      if (phase === mod.local.initial) {
        transitionLabel = deriveRootPhaseLabel(mod.moduleId);
      } else {
        const incoming = incomingMap.get(phase);
        if (incoming && incoming.length > 0) {
          const { obs, fromPhase } = incoming[0]!;
          const triggerCall = findTriggerCall(mod, fromPhase, obs);
          const fromGroup = phaseMap.get(fromPhase);
          const sourceTurn = fromGroup?.turn ?? null;
          transitionLabel = formatTransitionLabel(obs, triggerCall, sourceTurn);
        }
      }
    }

    result.push({
      phase,
      phaseDisplay: formatPhaseDisplay(phase, group.turn),
      turn: group.turn,
      transitionLabel,
      surfaces: group.surfaces,
    });
  }

  return result;
}

File: src-tauri/crates/bridge-session/src/session/learning_viewport.rs
//! Learning viewport builders — projects convention module internals into
//! viewport response types for UI consumption.
//!
//! Type definitions live in `learning_types.rs`. Formatting/text helpers
//! live in `learning_formatters.rs`. This file contains only the viewport
//! builder functions and their supporting helpers.
//!
//! Ported from TS `src/session/learning-viewport.ts`.

use std::collections::{HashMap, HashSet, VecDeque};

use bridge_conventions::pipeline::observation::normalize_intent::normalize_intent;
use bridge_conventions::pipeline::observation::route_matcher::match_obs;
use bridge_conventions::registry::bundle_registry::list_bundle_inputs;
use bridge_conventions::registry::module_registry::{
    get_all_modules, get_base_module_ids, get_module,
};
use bridge_conventions::rule_types::TurnRole;
use bridge_conventions::{BaseSystemId, ConventionModule, LocalFsm, ObsPattern, PhaseRef};
use bridge_engine::types::{BidSuit, Call};

use super::build_viewport::format_call;
use super::format_obs_label::format_transition_label;
use super::learning_formatters::{
    find_explanation_text, format_bid_references, format_module_name, format_phase_display,
    map_clauses, module_surfaces,
};
use super::learning_types::{
    BaseModuleInfo, EntryCondition, LearningTeachingView, ModuleCatalogEntry,
    ModuleLearningViewport, PhaseGroupView, RelevantMetric, ServiceTeachingLabel,
    SurfaceDetailView,
};

// ── PhaseRef helpers ─────────────────────────────────────────────────

fn phase_ref_to_vec(pr: &PhaseRef) -> Vec<&str> {
    match pr {
        PhaseRef::Single(s) => vec![s.as_str()],
        PhaseRef::Multiple(v) => v.iter().map(|s| s.as_str()).collect(),
    }
}

// ── Module catalog ───────────────────────────────────────────────────

/// Build module catalog entries for all registered modules.
pub fn build_module_catalog(system: BaseSystemId) -> Vec<ModuleCatalogEntry> {
    let all_modules = get_all_modules(system);
    let bundle_inputs = list_bundle_inputs();

    // Build reverse map: moduleId → bundleIds that contain it
    let mut module_bundles: HashMap<&str, Vec<String>> = HashMap::new();
    for input in bundle_inputs {
        for member_id in &input.member_ids {
            module_bundles
                .entry(member_id.as_str())
                .or_default()
                .push(input.id.clone());
        }
    }

    all_modules
        .iter()
        .map(|m| ModuleCatalogEntry {
            module_id: m.module_id.clone(),
            display_name: format_module_name(&m.module_id),
            description: format_bid_references(m.description.as_str()),
            purpose: format_bid_references(m.purpose.as_str()),
            surface_count: module_surfaces(m).len(),
            bundle_ids: module_bundles
                .get(m.module_id.as_str())
                .cloned()
                .unwrap_or_default(),
        })
        .collect()
}

/// Build read-only metadata for base system modules (for settings display).
pub fn build_base_module_infos(base_system_id: BaseSystemId) -> Vec<BaseModuleInfo> {
    let ids = get_base_module_ids(base_system_id);
    ids.iter()
        .filter_map(|&id| {
            let m = get_module(id, base_system_id)?;
            Some(BaseModuleInfo {
                id: id.to_string(),
                display_name: format_module_name(id),
                description: m.description.as_str().to_string(),
            })
        })
        .collect()
}

// ── Module learning viewport ─────────────────────────────────────────

/// Build a full learning viewport for a single module.
pub fn build_module_learning_viewport(
    module_id: &str,
    system: BaseSystemId,
) -> Option<ModuleLearningViewport> {
    let m = get_module(module_id, system)?;

    let bundle_inputs = list_bundle_inputs();
    let bundle_ids: Vec<String> = bundle_inputs
        .iter()
        .filter(|b| b.member_ids.iter().any(|mid| mid == module_id))
        .map(|b| b.id.clone())
        .collect();

    let teaching = &m.teaching;
    let phases = build_phase_groups(m, system);

    Some(ModuleLearningViewport {
        module_id: m.module_id.clone(),
        display_name: format_module_name(&m.module_id),
        description: format_bid_references(m.description.as_str()),
        purpose: format_bid_references(m.purpose.as_str()),
        teaching: LearningTeachingView {
            tradeoff: {
                let s = teaching.tradeoff.as_str();
                if s.is_empty() {
                    None
                } else {
                    Some(format_bid_references(s))
                }
            },
            principle: {
                let s = teaching.principle.as_str();
                if s.is_empty() {
                    None
                } else {
                    Some(format_bid_references(s))
                }
            },
            common_mistakes: teaching
                .common_mistakes
                .iter()
                .map(|item| format_bid_references(item.as_str()))
                .collect(),
        },
        phases,
        bundle_ids,
    })
}

// ── Phase ordering ───────────────────────────────────────────────────

/// Derive topological phase order from LocalFsm transitions via BFS.
pub fn derive_phase_order(fsm: &LocalFsm) -> Vec<String> {
    let mut phases = vec![fsm.initial.clone()];
    let mut seen = HashSet::new();
    seen.insert(fsm.initial.as_str().to_string());

    // Build adjacency map
    let mut adjacency: HashMap<String, Vec<String>> = HashMap::new();
    for t in &fsm.transitions {
        let froms = phase_ref_to_vec(&t.from);
        for f in froms {
            let entry = adjacency.entry(f.to_string()).or_default();
            if !entry.contains(&t.to) {
                entry.push(t.to.clone());
            }
        }
    }

    let mut queue = VecDeque::new();
    queue.push_back(fsm.initial.clone());

    while let Some(current) = queue.pop_front() {
        if let Some(neighbors) = adjacency.get(&current) {
            for next in neighbors {
                if seen.insert(next.clone()) {
                    phases.push(next.clone());
                    queue.push_back(next.clone());
                }
            }
        }
    }

    phases
}

/// Compute the set of post-fit phases for a module.
/// A phase is post-fit if any StateEntry at that phase has `negotiationDelta.fitAgreed`
/// truthy, or if it's reachable downstream from such a phase via FSM transitions.
pub fn compute_post_fit_phases(module: &ConventionModule) -> HashSet<String> {
    let mut fit_phases = HashSet::new();

    for entry in module.states.as_deref().unwrap_or(&[]) {
        let has_fit = entry
            .negotiation_delta
            .as_ref()
            .and_then(|nd| nd.fit_agreed.as_ref())
            .is_some();

        if has_fit {
            for p in phase_ref_to_vec(&entry.phase) {
                fit_phases.insert(p.to_string());
            }
        }
    }

    // Build adjacency from transitions
    let mut adjacency: HashMap<String, Vec<String>> = HashMap::new();
    for t in &module.local.transitions {
        for f in phase_ref_to_vec(&t.from) {
            let entry = adjacency.entry(f.to_string()).or_default();
            if !entry.contains(&t.to) {
                entry.push(t.to.clone());
            }
        }
    }

    // BFS from fit-establishing phases
    let mut result = fit_phases.clone();
    let mut queue: VecDeque<String> = fit_phases.into_iter().collect();
    while let Some(current) = queue.pop_front() {
        if let Some(neighbors) = adjacency.get(&current) {
            for next in neighbors {
                if result.insert(next.clone()) {
                    queue.push_back(next.clone());
                }
            }
        }
    }

    result
}

// ── Entry condition ──────────────────────────────────────────────────

/// Capability → entry condition mapping.
fn cap_entry_conditions() -> HashMap<&'static str, EntryCondition> {
    let mut map = HashMap::new();
    map.insert(
        "opening.1nt",
        EntryCondition {
            label: "Partner opened 1NT".to_string(),
            call: Some(Call::Bid {
                level: 1,
                strain: BidSuit::NoTrump,
            }),
            turn: Some("opener".to_string()),
        },
    );
    map.insert(
        "opening.major",
        EntryCondition {
            label: "Partner opened a major".to_string(),
            call: None,
            turn: Some("opener".to_string()),
        },
    );
    map.insert(
        "opening.weak-two",
        EntryCondition {
            label: "Partner opened a weak two".to_string(),
            call: None,
            turn: Some("opener".to_string()),
        },
    );
    map.insert(
        "opponent.1nt",
        EntryCondition {
            label: "Opponent opened 1NT".to_string(),
            call: Some(Call::Bid {
                level: 1,
                strain: BidSuit::NoTrump,
            }),
            turn: None,
        },
    );
    map
}

/// Get the primary capability key from declared capabilities.
fn get_primary_capability(
    declared_capabilities: Option<&HashMap<String, String>>,
) -> Option<String> {
    let caps = declared_capabilities?;
    caps.keys().next().cloned()
}

/// Derive full entry condition from the module's host capability.
pub fn derive_entry_condition(module_id: &str) -> Option<EntryCondition> {
    let conditions = cap_entry_conditions();
    for input in list_bundle_inputs() {
        if !input.member_ids.iter().any(|id| id == module_id) {
            continue;
        }
        let cap_id = get_primary_capability(input.declared_capabilities.as_ref())?;
        if let Some(ec) = conditions.get(cap_id.as_str()) {
            return Some(ec.clone());
        }
    }
    None
}

/// Derive root phase label from the module's host capability.
fn derive_root_phase_label(module_id: &str) -> Option<String> {
    derive_entry_condition(module_id).map(|ec| ec.label)
}

/// Find the first surface at a given phase whose normalized intent matches a transition obs pattern.
fn find_trigger_call(
    module: &ConventionModule,
    from_phase: &str,
    obs: &ObsPattern,
) -> Option<Call> {
    for entry in module.states.as_deref().unwrap_or(&[]) {
        let entry_phases = phase_ref_to_vec(&entry.phase);
        if !entry_phases.contains(&from_phase) {
            continue;
        }
        for surface in &entry.surfaces {
            let actions = normalize_intent(&surface.source_intent);
            if actions.iter().any(|a| match_obs(obs, a, None)) {
                return Some(surface.encoding.default_call.clone());
            }
        }
    }
    None
}

/// Convert TurnRole to display string.
fn turn_role_display(role: TurnRole) -> &'static str {
    match role {
        TurnRole::Opener => "opener",
        TurnRole::Responder => "responder",
        TurnRole::Opponent => "opponent",
    }
}

/// Build PhaseGroupView[] from a module's states, ordered by FSM topology.
fn build_phase_groups(module: &ConventionModule, _system: BaseSystemId) -> Vec<PhaseGroupView> {
    let states = module.states.as_deref().unwrap_or(&[]);
    if states.is_empty() {
        return Vec::new();
    }

    let phase_order = derive_phase_order(&module.local);
    let post_fit_phases = compute_post_fit_phases(module);

    // Build incoming transition map
    let mut incoming_map: HashMap<String, Vec<(ObsPattern, String)>> = HashMap::new();
    for t in &module.local.transitions {
        for f in phase_ref_to_vec(&t.from) {
            incoming_map
                .entry(t.to.clone())
                .or_default()
                .push((t.on.clone(), f.to_string()));
        }
    }

    // Group states by phase string (flatten multi-phase entries)
    struct PhaseGroup {
        turn: Option<TurnRole>,
        surfaces: Vec<SurfaceDetailView>,
    }

    let mut phase_map: HashMap<String, PhaseGroup> = HashMap::new();

    for entry in states {
        let entry_phases = phase_ref_to_vec(&entry.phase);

        for phase in entry_phases {
            let metric = if post_fit_phases.contains(phase) {
                Some(RelevantMetric::TrumpTp)
            } else {
                Some(RelevantMetric::Hcp)
            };

            let group = phase_map
                .entry(phase.to_string())
                .or_insert_with(|| PhaseGroup {
                    turn: entry.turn,
                    surfaces: Vec::new(),
                });

            let seen: HashSet<String> = group
                .surfaces
                .iter()
                .map(|s| s.meaning_id.clone())
                .collect();

            for surface in &entry.surfaces {
                if seen.contains(&surface.meaning_id) {
                    continue;
                }

                let raw_explanation =
                    find_explanation_text(&module.explanation_entries, &surface.meaning_id);

                group.surfaces.push(SurfaceDetailView {
                    meaning_id: surface.meaning_id.clone(),
                    teaching_label: ServiceTeachingLabel {
                        name: format_bid_references(surface.teaching_label.name.as_str()),
                        summary: format_bid_references(surface.teaching_label.summary.as_str()),
                    },
                    call: surface.encoding.default_call.clone(),
                    call_display: format_call(&surface.encoding.default_call),
                    disclosure: surface.disclosure,
                    recommendation: Some(surface.ranking.recommendation_band),
                    explanation_text: raw_explanation.map(|t| format_bid_references(&t)),
                    clauses: map_clauses(&surface.clauses, metric),
                });
            }
        }
    }

    // Determine visible phases
    let visible_phases: Vec<&str> = phase_order
        .iter()
        .filter(|p| {
            phase_map
                .get(p.as_str())
                .map_or(false, |g| !g.surfaces.is_empty())
        })
        .map(|p| p.as_str())
        .collect();
    let suppress_labels = visible_phases.len() < 3;

    let mut result = Vec::new();
    for phase in &phase_order {
        let group = match phase_map.get(phase.as_str()) {
            Some(g) if !g.surfaces.is_empty() => g,
            _ => continue,
        };

        let transition_label = if suppress_labels {
            None
        } else if phase == &module.local.initial {
            derive_root_phase_label(&module.module_id)
        } else {
            incoming_map.get(phase.as_str()).and_then(|incoming| {
                let (obs, from_phase) = incoming.first()?;
                let trigger_call = find_trigger_call(module, from_phase, obs);
                let from_group = phase_map.get(from_phase.as_str());
                let source_turn = from_group.and_then(|g| g.turn);
                let turn_str = source_turn.map(turn_role_display);
                Some(format_transition_label(
                    obs,
                    trigger_call.as_ref(),
                    turn_str,
                ))
            })
        };

        let turn_str = group.turn.map(turn_role_display);
        result.push(PhaseGroupView {
            phase: phase.clone(),
            phase_display: format_phase_display(phase, turn_str),
            turn: turn_str.map(|s| s.to_string()),
            transition_label,
            surfaces: group.surfaces.clone(),
        });
    }

    result
}

// ── Tests ────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use bridge_conventions::{BidActionType, LocalFsm, ObsPatternAct, PhaseTransition};

    #[test]
    fn build_module_catalog_returns_10() {
        let catalog = build_module_catalog(BaseSystemId::Sayc);
        assert_eq!(catalog.len(), 10);
    }

    #[test]
    fn build_module_catalog_has_stayman() {
        let catalog = build_module_catalog(BaseSystemId::Sayc);
        let stayman = catalog.iter().find(|e| e.module_id == "stayman");
        assert!(stayman.is_some());
        let s = stayman.unwrap();
        assert_eq!(s.display_name, "Stayman");
        assert!(s.surface_count > 0);
    }

    #[test]
    fn build_module_learning_viewport_stayman() {
        let viewport = build_module_learning_viewport("stayman", BaseSystemId::Sayc);
        assert!(viewport.is_some());
        let v = viewport.unwrap();
        assert_eq!(v.module_id, "stayman");
        assert_eq!(v.display_name, "Stayman");
        assert!(!v.phases.is_empty());
    }

    #[test]
    fn build_module_learning_viewport_unknown() {
        let viewport = build_module_learning_viewport("nonexistent", BaseSystemId::Sayc);
        assert!(viewport.is_none());
    }

    #[test]
    fn derive_phase_order_simple_fsm() {
        let fsm = LocalFsm {
            initial: "idle".to_string(),
            transitions: vec![
                PhaseTransition {
                    from: PhaseRef::Single("idle".to_string()),
                    to: "asked".to_string(),
                    on: ObsPattern {
                        act: ObsPatternAct::Specific(BidActionType::Inquire),
                        feature: None,
                        suit: None,
                        strain: None,
                        strength: None,
                        actor: None,
                    },
                },
                PhaseTransition {
                    from: PhaseRef::Single("asked".to_string()),
                    to: "responded".to_string(),
                    on: ObsPattern {
                        act: ObsPatternAct::Specific(BidActionType::Show),
                        feature: None,
                        suit: None,
                        strain: None,
                        strength: None,
                        actor: None,
                    },
                },
            ],
        };
        let order = derive_phase_order(&fsm);
        assert_eq!(order, vec!["idle", "asked", "responded"]);
    }

    #[test]
    fn derive_phase_order_multi_from() {
        let fsm = LocalFsm {
            initial: "a".to_string(),
            transitions: vec![
                PhaseTransition {
                    from: PhaseRef::Single("a".to_string()),
                    to: "b".to_string(),
                    on: ObsPattern {
                        act: ObsPatternAct::Any,
                        feature: None,
                        suit: None,
                        strain: None,
                        strength: None,
                        actor: None,
                    },
                },
                PhaseTransition {
                    from: PhaseRef::Multiple(vec!["a".to_string(), "b".to_string()]),
                    to: "c".to_string(),
                    on: ObsPattern {
                        act: ObsPatternAct::Any,
                        feature: None,
                        suit: None,
                        strain: None,
                        strength: None,
                        actor: None,
                    },
                },
            ],
        };
        let order = derive_phase_order(&fsm);
        assert_eq!(order, vec!["a", "b", "c"]);
    }

    #[test]
    fn build_base_module_infos_returns_4() {
        let infos = build_base_module_infos(BaseSystemId::Sayc);
        assert_eq!(infos.len(), 4);
    }
}
